File: python/vigenere.py
import string

chars = string.ascii_uppercase + string.ascii_lowercase

def encrypt(text, key):

    mappedText = [chars.find(i) for i in text]
    mappedKey = [chars.find(i) for i in key]

    keyLength = len(key)
    textLength = len(text)

    cipher = ""

    for n in range(textLength):
        letter = (mappedText[n] + mappedKey[n % keyLength]) % len(chars)
        cipher += chars[letter]

    return cipher

def decrypt(cipher, key):
    
    mappedCipher = [chars.find(i) for i in cipher]
    mappedKey = [chars.find(i) for i in key]

    keyLength = len(key)
    cipherLength = len(cipher)

    plainText = ""

    for n in range(cipherLength):
        letter = (mappedCipher[n] - mappedKey[n % keyLength]) % len(chars)
        plainText += chars[letter]

    return plainText


if __name__ == "__main__":
    
    text = input('Text: ')
    key = input('Key: ')

    cipher = encrypt(text, key)
    plainText = decrypt(cipher, key)

    print('=======================')
    print('Text: ', text)
    print('Key: ', key)
    print('Cipher: ', cipher)
    print('Plain Text: ', plainText)

File: c++/vigenere.cpp
#include "vigenere.hpp"

Vigenere::Vigenere() {}
Vigenere::~Vigenere() {}

vector<size_t> Vigenere::getCharPositions(string& str) {

    vector<size_t> charsAtPos;
    for (int i = 0; i < str.length(); i++) {
        charsAtPos.push_back(chars.find(str.at(i)));
    }

    return charsAtPos;
}

string Vigenere::encrypt(string text, string key) {

    vector<size_t> textCharAtPos = getCharPositions(text);
    vector<size_t> keyCharAtPos = getCharPositions(key);

    size_t textLength = text.length();
    size_t keyLength = key.length();

    string cipher;
    int letter;

    for (int n = 0; n < textLength; n++) {
        letter = (textCharAtPos.at(n) + keyCharAtPos.at(n % keyLength));
        letter %= chars.length();
        cipher += chars.at(letter);
    }

    return cipher;
}

string Vigenere::decrypt(string cipher, string key) {

    vector<size_t> cipherCharAtPos = getCharPositions(cipher);
    vector<size_t> keyCharAtPos = getCharPositions(key);

    size_t cipherLength = cipher.length();
    size_t keyLength = key.length();

    string plainText;
    int letter;

    for (int n = 0; n < cipherLength; n++) {
        letter = (cipherCharAtPos[n] - keyCharAtPos[n % keyLength]);
        letter += chars.length();
        letter %= chars.length();
        plainText += chars.at(letter);
    }

    return plainText;
}